src/main/java/com/lgtm/easymoney/services/impl/LoanServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+import org.springframework.stereotype.Service;
+
+/**
+ * Loan service implementation.
+ */
+@Service("loanService")
+public class LoanServiceImpl implements LoanService {
+  private final TransactionService transactionService;
+  private final UserService userService;
+  private final RequestService requestService;
+
+  /**
+   * Loan service implementation.
+   */
+  @Autowired
+  public LoanServiceImpl(
+      TransactionService transactionService,
+      UserService userService,
+      RequestService requestService) {
+    this.transactionService = transactionService;
+    this.userService = userService;
+    this.requestService = requestService;
+  }
+
+  private void validateLoanUsers(User borrower, User lender) {
+    if (!lender.getType().equals(UserType.FINANCIAL)) {
+      throw new InapplicableOperationException("user", lender.getId(), "toUid", "requestLoan");
+    }
+    if (!borrower.getType().equals(UserType.PERSONAL)) {
+      throw new InapplicableOperationException("user", borrower.getId(), "fromUid", "requestLoan");
+    }
+  }
+
+  private Transaction createLoanRequest(User reqBy, User reqTo, BigDecimal amount, String desc,
+      Category category) {
+    Transaction trans = new Transaction();
+    // request: requested by A, requested to B
+    // once accept, money goes from B to A
+    trans.setFrom(reqTo);
+    trans.setTo(reqBy);
+    trans.setAmount(amount);
+    trans.setDescription(desc);
+    trans.setCategory(category);
+    trans.setStatus(TransactionStatus.LOAN_PENDING);
+    return transactionService.saveTransaction(trans);
+  }
+
+  private List<Transaction> getLoansByUser(User user) {
+    List<TransactionStatus> status = List.of(
+        TransactionStatus.LOAN_PENDING,
+        TransactionStatus.LOAN_APPROVED,
+        TransactionStatus.LOAN_DECLINED);
+    return transactionService.getAllTransactionsWithUser(user, status);
+  }
+
+  private Transaction getLoanById(Long id) {
+    return transactionService.getTransactionById(id);
+  }
+
+  private boolean validateLoanRequest(Long lid, Long fromUid, Long toUid) {
+    Transaction t = getLoanById(lid);
+    return t.getFrom().getId().equals(fromUid)
+        && t.getTo().getId().equals(toUid)
+        && t.getStatus() == TransactionStatus.LOAN_PENDING;
+  }
+
+  @Override
+  public ResourceCreatedRsp requestLoan(RequestReq req) {
+    // get params
+    BigDecimal amount = req.getAmount();
+    String category = req.getCategory();
+    String desc = req.getDescription();
+    User borrower = userService.getUserById(req.getFromUid());
+    User lender = userService.getUserById(req.getToUid());
+    // validate if loan user type is valid
+    validateLoanUsers(borrower, lender);
+    // create loan request
+    Transaction loan = createLoanRequest(borrower, lender, amount, desc,
+        Category.valueOf(category.toUpperCase()));
+    // response
+    return loan == null ? new ResourceCreatedRsp(null) : new ResourceCreatedRsp(loan.getId());
```

Comment:
repo.save() never returns null. If there's issues saving, it will throw some built-in exception like DataAccessException, so no need to check null

Reply:
yeah i think transaction need to remove this too

Reply:
Modified for both loanServiceImple and requestServiceImpl. Will push the change together with next commit.